lcls-twincat-common-components/Library/Devices/PPM/FB_PPM_PowerMeter.TcPOU

```diff
@@ -99,19 +160,91 @@ fbGetPMVoltage(
     fTermMin := 0,
     fReal => fVoltage);
 
-// Power meter calibration
-fCalibBase := (fVoltage + fCalibRelOffset) * fCalibRelRatio;
-fCalibMJ := fCalibBase * fCalibMJRatio;
+IF R_TRIG(CLK:=(eEnumGet <> E_PPM_States.OUT OR stYAxisState <> 0)) THEN
+    // Dump most recent readings
+    IF udBackgroundVoltageBufferIndex > 5 THEN
+        FOR i := (udBackgroundVoltageBufferIndex - 5) TO udBackgroundVoltageBufferIndex DO
+            fBackgroundVoltageSum := fBackgroundVoltageSum - fBackgroundVoltageBuffer[i MOD 100];
+        END_FOR
+        udBackgroundVoltageBufferIndex := udBackgroundVoltageBufferIndex - 5;
+        IF udBackgroundVoltageBufferIndex <> 1 AND BACKGROUND_MODE = Passive THEN
+            fBackgroundVoltage := fBackgroundVoltageSum / MAX((udBackgroundVoltageBufferIndex - 1), 100);
+        END_IF
+    END_IF
+    fBackgroundVoltageAcc := 0;
+    fBackgroundVoltageSum := 0;
+    tonBackgroundAutoCollecting(IN := FALSE);
+    udBackgroundVoltageBufferIndex := 1;
+ELSIF eEnumGet = E_PPM_States.OUT AND stYAxisState = 0 THEN
+    tonBackgroundAutoCollecting( IN := TRUE, PT := T#1000MS);
+    IF tonBackgroundAutoCollecting.Q = TRUE THEN
+        tonBackgroundAutoCollecting( IN := FALSE);
+        fBackgroundVoltageStale := fBackgroundVoltageBuffer[udBackgroundVoltageBufferIndex];
+        fBackgroundVoltageBuffer[udBackgroundVoltageBufferIndex] := fBackgroundVoltageAcc;
+        IF udBackgroundVoltageBufferIndex > 100 THEN
+            fBackgroundVoltageSum := fBackgroundVoltageSum - fBackgroundVoltageStale;
+        END_IF
+        fBackgroundVoltageSum := fBackgroundVoltageSum + fBackgroundVoltageAcc;
```

Comment:
Is it possible for these voltage sums to accumulate floating point addition errors if we keep subtracting from and adding to them like this?

Reply:
There will be some error, but I don't expect it to be large enough to be significant. The alternative I know of is to sume over the array every cycle (or every n cycles), but it would make those cycles significantly longer.

Reply:
You're probably right. We should give this a try and see if there's any other nuances.

Reply:
Wanted to briefly revisit this. Let's say we were worried about floating-point error accumulation, and decided to sum over a hundred-element array every second. How would you quantify the increase in cycle time this creates for the cycle where that sum occurs and the likelihood of exceeding the cycle time for a program that has ppm function blocks?